File: src/main/resources/assets/bonsaitrees/config/types.d/pamsharvestcraft.js
load("config/bonsaitrees/types.d/defaults.js");

isEnabled = function() {
    var Loader = Java.type("net.minecraftforge.fml.common.Loader");
    return Loader.isModLoaded("harvestcraft");
}

var main = function(source) {
    var FruitRegistry = Java.type("com.pam.harvestcraft.blocks.FruitRegistry");
    var BlockPamSapling = Java.type("com.pam.harvestcraft.blocks.growables.BlockPamSapling");

    var ItemStack = Java.type("net.minecraft.item.ItemStack");
    var PamHelper = Java.type("org.dave.bonsaitrees.compat.PamHelper");

    FruitRegistry.getSaplings().forEach(function(sapling) {
        var name = sapling.getRegistryName().toString().replace("_sapling", "");

        var pamFruitBlock = sapling.getFruit();
        var pamFruitItem = pamFruitBlock.getFruitItem();

        var pamTreeType = new TreeTypeSimple(name, new ItemStack(sapling, 1, 0));
        pamTreeType.setSource(source);
        pamTreeType.addDrop("minecraft:stick", 0, stickAmount, stickChance);
        pamTreeType.addDrop(new ItemStack(sapling, saplingAmount, 0), saplingChance);

        var logState = PamHelper.reflectPamSapling(sapling, "logState");
        pamTreeType.addDrop(new ItemStack(logState.getBlock(), logAmount, logState.getBlock().getMetaFromState(logState)), logChance);

        pamTreeType.addDrop(new ItemStack(pamFruitItem, fruitAmount, 0), fruitChance);

        var leafState = PamHelper.reflectPamSapling(sapling, "leavesState");
        pamTreeType.addDrop(new ItemStack(leafState.getBlock(), leafAmount, leafState.getBlock().getMetaFromState(leafState)), leafChance);

        TreeTypeRegistry.registerTreeType(pamTreeType);
    });
};

var generateTree = function(type, world, pos, rand) {
    var FruitRegistry = Java.type("com.pam.harvestcraft.blocks.FruitRegistry");
    var BlockPamSapling = Java.type("com.pam.harvestcraft.blocks.growables.BlockPamSapling");

    FruitRegistry.getSaplings().forEach(function(sapling) {
        var name = sapling.getRegistryName().toString().replace("_sapling", "");
        if(name != type.typeName) {
            return;
        }

        world.setBlockState(pos, sapling.getDefaultState());
        sapling.grow(world, rand, pos, sapling.getDefaultState());
    });
};

File: src/main/java/org/dave/bonsaitrees/integration/mods/PamsHarvestcraft.java
package org.dave.bonsaitrees.integration.mods;

import com.pam.harvestcraft.blocks.FruitRegistry;
import com.pam.harvestcraft.blocks.growables.BlockPamSapling;
import net.minecraft.block.state.IBlockState;
import net.minecraft.init.Items;
import net.minecraft.item.ItemStack;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.ReflectionHelper;
import org.dave.bonsaitrees.api.*;

import java.util.Random;

@BonsaiIntegration(mod = "harvestcraft")
public class PamsHarvestcraft implements IBonsaiIntegration {
    @Override
    public void registerTrees(ITreeTypeRegistry registry) {
        for(BlockPamSapling sapling : FruitRegistry.getSaplings()) {
            String name = sapling.getRegistryName().toString().replace("_sapling", "");

            TreeTypeSimple pamTreeType = new TreeTypeSimple(name, new ItemStack(sapling, 1, 0));

            pamTreeType.addDrop(new ItemStack(sapling, BonsaiDropChances.saplingAmount), BonsaiDropChances.saplingChance);
            pamTreeType.addDrop(new ItemStack(Items.STICK, BonsaiDropChances.stickAmount), BonsaiDropChances.stickChance);
            pamTreeType.addDrop(new ItemStack(sapling.getFruitItem(), BonsaiDropChances.fruitAmount), BonsaiDropChances.fruitChance);

            IBlockState logState = ReflectionHelper.getPrivateValue(BlockPamSapling.class, sapling, "logState");
            pamTreeType.addDrop(new ItemStack(logState.getBlock(), BonsaiDropChances.logAmount), BonsaiDropChances.logChance);

            IBlockState leafState = ReflectionHelper.getPrivateValue(BlockPamSapling.class, sapling, "leavesState");
            pamTreeType.addDrop(new ItemStack(leafState.getBlock(), BonsaiDropChances.leafAmount), BonsaiDropChances.leafChance);

            registry.registerTreeType(this, pamTreeType);
        }
    }

    @Override
    public void generateTree(IBonsaiTreeType type, World world, BlockPos pos, Random rand) {
        for(BlockPamSapling sapling : FruitRegistry.getSaplings()) {
            String name = sapling.getRegistryName().toString().replace("_sapling", "");
            if(!name.equals(type.getName())) {
                continue;
            }

            world.setBlockState(pos, sapling.getDefaultState());
            sapling.grow(world, rand, pos, sapling.getDefaultState());
        }
    }
}
